LightBulb/ViewModels/Components/DashboardViewModel.cs

```diff
@@ -285,12 +288,63 @@ private void UpdateInstant()
         }
     }
 
-    private void UpdateConfiguration()
+    private const double _brightnessDefaultStep = 0.08;
+    private const double _temperatureDefaultStep = 30;
+
+    private double _brightnessMaxStep = _brightnessDefaultStep;
+    private double _temperatureMaxStep = _temperatureDefaultStep;
+
+    private ColorConfiguration _lastTarget;
+
+    private void UpdateConfiguration(double stepsPerSecond)
     {
-        var isSmooth = _settingsService.IsConfigurationSmoothingEnabled && !IsCyclePreviewEnabled;
+        if (CurrentConfiguration == TargetConfiguration)
+            return;
+
+        var isSmooth =
+            _settingsService.IsConfigurationSmoothingEnabled
+            && !IsCyclePreviewEnabled
+            && _settingsService.MaxSettingsTransitionDuration.TotalSeconds >= 0.1;
+
+        // If we've changed targets, restart with default settings.
+        if (_lastTarget != TargetConfiguration && isSmooth)
+        {
+            _brightnessMaxStep = _brightnessDefaultStep;
+            _temperatureMaxStep = _temperatureDefaultStep;
+            _lastTarget = TargetConfiguration;
+
+            var tempDelta = Math.Abs(
+                TargetConfiguration.Temperature - CurrentConfiguration.Temperature
+            );
+            var brightnessDelta = Math.Abs(
+                TargetConfiguration.Brightness - CurrentConfiguration.Brightness
+            );
+            var expectedTemperatureDuration = tempDelta / (_temperatureMaxStep * stepsPerSecond);
+            var expectedBrightnessDuration =
+                brightnessDelta / (_brightnessMaxStep * stepsPerSecond);
+
+            // If the expected durations are longer than our duration limit, we adjust the step amount to stay at the max duration.
+            var goalDuration = Math.Max(expectedTemperatureDuration, expectedBrightnessDuration);
+            goalDuration = Math.Min(
+                goalDuration,
+                _settingsService.MaxSettingsTransitionDuration.TotalSeconds
+            );
+
+            // Calculate the step-rate needed to reach the goal.
+            _brightnessMaxStep = brightnessDelta / (goalDuration * stepsPerSecond);
+            _temperatureMaxStep = tempDelta / (goalDuration * stepsPerSecond);
+
+            // If we ended up slower on either of the durations, speed us up.
+            _brightnessMaxStep = Math.Max(_brightnessMaxStep, _brightnessDefaultStep);
+            _temperatureMaxStep = Math.Max(_temperatureMaxStep, _temperatureDefaultStep);
+        }
 
         CurrentConfiguration = isSmooth
-            ? CurrentConfiguration.StepTo(TargetConfiguration, 30, 0.008)
+            ? CurrentConfiguration.StepTo(
+                TargetConfiguration,
+                _temperatureMaxStep,
+                _brightnessMaxStep
+            )
             : TargetConfiguration;
```

Comment:
What about instead of configuring smoothing duration, you could configure the steps instead? You'd then be able to set arbitrary values instead of `30` and `0.008`, which would let you speed up the transitions, but the code would end up a lot simpler (basically same as before).

Reply:
I tried that, but it ended up leading to strobing in the case of small transitions at high speed. If a small configuration change was made, such as sliding a slider, it led to instant changes similar to gamma smoothing being disabled entirely. But if that's preferable behavior, I can change it to that.